### ISSUE TITLE
gnupg21: gpg2 --gen-key error "No pinentry"

### ISSUE THREAD

**Issue by bzz** [Homebrew/homebrew-versions#681]
As suggested in Homebrew/homebrew#12382 opening a new issue here.

Not able to generate a key on osx 10.10.2, looks alike with http://sourceforge.net/p/gpgosx/bugs/1/

```
gpg2 --debug-level advanced --expert --full-gen-key

pg: DBG: chan_4 <- OK Pleased to meet you
gpg: DBG: connection to agent established
gpg: DBG: chan_4 -> RESET
gpg: DBG: chan_4 <- OK
gpg: DBG: chan_4 -> OPTION ttyname=/dev/ttys001
gpg: DBG: chan_4 <- OK
gpg: DBG: chan_4 -> OPTION ttytype=xterm-256color
gpg: DBG: chan_4 <- OK
gpg: DBG: chan_4 -> OPTION lc-ctype=UTF-8
gpg: DBG: chan_4 <- OK
gpg: DBG: chan_4 -> OPTION lc-messages=en_US.UTF-8
gpg: DBG: chan_4 <- OK
gpg: DBG: chan_4 -> OPTION allow-pinentry-notify
gpg: DBG: chan_4 <- OK
gpg: DBG: chan_4 -> OPTION agent-awareness=2.1.0
gpg: DBG: chan_4 <- OK
gpg: DBG: chan_4 -> AGENT_ID
gpg: DBG: chan_4 <- ERR 67109139 Unknown IPC command <GPG Agent>
gpg: DBG: chan_4 -> RESET
gpg: DBG: chan_4 <- OK
gpg: DBG: chan_4 -> GENKEY
gpg: DBG: chan_4 <- S INQUIRE_MAXLEN 1024
gpg: DBG: chan_4 <- INQUIRE KEYPARAM
gpg: DBG: chan_4 -> D (genkey(ecc(curve 10:NIST P-521)(flags nocomp)))
gpg: DBG: chan_4 -> END
gpg: DBG: chan_4 <- ERR 67108949 No pinentry <GPG Agent>
gpg: agent_genkey failed: No pinentry
Key generation failed: No pinentry
gpg: random usage: poolsize=600 mixed=0 polls=0/0 added=0/0
              outmix=0 getlvl1=0/0 getlvl2=0/0
gpg: secmem usage: 1344/32768 bytes in 2 blocks


less ~/.gnupg/gpg-agent.conf                                                                          ⮂
/Users/alex/.gnupg/gpg-agent.conf: No such file or directory


brew --config                                                                                         ⮂
HOMEBREW_VERSION: 0.9.5
ORIGIN: https://github.com/Homebrew/homebrew
HEAD: 7e03127543deb7a1b5646d1cafa342af3cf4e07f
Last commit: 6 hours ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CELLAR: /usr/local/Cellar
CPU: quad-core 64-bit haswell
OS X: 10.10.2-x86_64
Xcode: 6.1.1
CLT: 6.1.1.0.1.1416017670
Clang: 6.0 build 600
X11: 2.7.7 => /opt/X11
System Ruby: 2.0.0-p481
Perl: /usr/bin/perl
Python: /usr/local/bin/python => /usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/bin/python2.7
Ruby: /usr/bin/ruby
Java: 1.7.0_45
```
The same happens after adding existing pinentry to ```~/.gnupg/gpg-agent.conf``` 
```
pinentry-program /usr/local/bin/pinentry
```

**Comment by DomT4** [comment]
Hmm. This is a little odd. Can you try ` brew cleanup && brew prune ` and then try reinstalling ` gnupg21 ` and see if the problem replicates? Wondering if there's something old sticking around causing the issue. If that doesn't work, a ` brew doctor && brew config ` would be helpful.

**Comment by bzz** [comment]
Thanks for helping with this one, It is very odd indeed.
After cleanup, prune and reinstall (poured from bottle), got the same result with 

```shell
gpg: agent_genkey failed: No pinentry
Key generation failed: No pinentry
```

```
$brew doctor                                                     ⮂
Warning: Some installed formula are missing dependencies.
You should `brew install` the missing dependencies:

    brew install d-bus

Run `brew missing` for more details.

Warning: Some keg-only formula are linked into the Cellar.
Linking a keg-only formula, such as gettext, into the cellar with
`brew link <formula>` will cause other formulae to detect them during
the `./configure` step. This may cause problems when compiling those
other formulae.

Binaries provided by keg-only formulae may override system binaries
with other strange results.

You may wish to `brew unlink` these brews:

    scala210
```

**Comment by DomT4** [comment]
Hmm. I haven't even seen any reports like this on the GnuPG mailing lists recently, very odd.

What does ` brew list | grep pinentry ` cough out for you?

**Comment by bzz** [comment]
```
$brew list | grep pinentry                                                                                                                                 
pinentry

$which pinentry | xargs ls -l                                                                                                                           
lrwxr-xr-x 1 user admin 37 Feb 21 21:46 /usr/local/bin/pinentry -> ../Cellar/pinentry/0.9.0/bin/pinentry

$pinentry                                                                                                                                                 
OK Your orders please
````
And I tried to install ```pinentry-mac``` and putting it into ```~/.gnupg/gpg-agent.conf``` with exactly the same result - error on ```pgp2 --gen-key```.

I also tried installing gnupg21 on a different, clean os x 10.10.2 machine with no luck either - same error.

Would appreciate further directions as I'm running out of Ideas

**Comment by DomT4** [comment]
Our logs differ on the same command at this point:

```
gpg: DBG: chan_4 <- ERR 67109139 Unknown IPC command <GPG Agent>
gpg: DBG: chan_4 -> RESET
gpg: DBG: chan_4 <- OK
```

Mine shows:

```
gpg: DBG: chan_3 <- ERR 67109139 Unknown IPC command <GPG Agent>
gpg: DBG: chan_3 -> GETINFO version
gpg: DBG: chan_3 <- D 2.1.2
gpg: DBG: chan_3 <- OK
```

Can you try ` brew remove pinentry gnupg21 ` and then just ` brew install gnupg2 ` and then try again?

Make sure ` gnupg2 gpg-agent dirmngr fwknop gpgme ` are completely removed from your system as well, rather than just unlinked. Shouldn't be an issue, but with the no-compatibility between 2.x and 2.1.x better not to take chances.

**Comment by DomT4** [comment]
Potentially worth trying ` killall gpg-agent && gpg-agent --daemon --use-standard-socket --pinentry-program /usr/local/bin/pinentry ` as well.

**Comment by bzz** [comment]
@DomT4  thank you very much, killing a re-starting the agent manually did the trick.

Guess it is good to close this issue as the solution is documented (although I'm not sure how to avoid it for other users)

**Comment by DomT4** [comment]
Awesome. If we get a particularly large number of users complaining of the same thing I'll add a caveat instructing people that it's worthwhile killing the gpg-agent after install and then manually restarting it. Glad it's working for you now!

**Comment by vdm** [comment]
I had uninstalled `gpg`, `gnupg2` and `gnupg21`, but it still took restarting `gpg-agent` to fix this for me. Thank you @bzz and @DomT4.

A link to here should be on the caveats. It would also be helpful to have a link to this issue from https://github.com/Homebrew/homebrew/issues/12382, but it has been locked by @tdsmith.

**Comment by DomT4** [comment]
I'll add a caveat to the formula on the ` gpg-agent ` issue shortly.

**Comment by DomT4** [comment]
https://github.com/Homebrew/homebrew-versions/commit/053881fff245dd5d65e12307f3492db5887ceea1

**Comment by Sidnicious** [comment]
FWIW it's only working for me if I start the agent manually with

    gpg-agent --daemon --pinentry-program /usr/local/bin/pinentry

**Comment by ay** [comment]
@Sidnicious Yeah, I just came across this too. I agree that it should just work and I've opened a pull request to fix this at #983.

**Comment by KonVas** [comment]
I wonder if this is related, but is the only closest thread I could find to reach for a solution, alas, I still can't fix it. I am getting this response in Emacs 
"Error while decrypting with "/usr/local/bin/gpg1":
gig: Invalid option "--pinentry-mode" "
